fuzz: improve utxo_snapshot target

mzumsande opened the pull request:
Add the possibility of giving more guidance to the creation of the metadata and/or coins, so that the fuzzer gets the chance
to reach more error conditions in ActivateSnapshot and sometimes successfully creates a valid snapshot.

This also changes the asserts for the success case that were outdated (after #29370) and only didn't result in a crash because the fuzzer wasn't able to reach this code before.

mzumsande committed:
fuzz: improve utxo_snapshot target

Add the possibility of giving more guidance to the creation of the
metadata and/or coins, so that the fuzzer gets the chance
to reach more error conditions in ActivateSnapshot and sometimes
successfully creates a valid snapshot.

This also changes the asserts for the success case that were outdated,
and only didn't result in a crash because the fuzzer wasn't able
to reach this code before.

fjahr commented:
Concept ACK

maflcko reviewed (approved):
lgtm. Thanks!

maflcko reviewed (commented):
ACK b9ba1a73094f4ad593b527e23d2681f41c22539 🎯

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK b9ba1a73094f4ad593b527e23d2681f41c22539 🎯
iTX6VgeBMzgsveHFGIl5Qnsj/cWt0KYcv2B3dvhxmLm6ljOkWyT0bGPRv0zKVRDXMpcDBJcwelxlW8o6WHPEDw==
```

</details>

TheCharlatan commented:
I checked if there is anything we could do in the testing setup to make the test faster. With this patchset, the test takes around 1:15 min for 10k runs on my machine. Removing the validation interface from the test setup cuts it down to 1:05 min , if I remove the networking setup, it takes 35 sec to complete.

I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?

fjahr commented:
> I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?

That appears to be a bug. We actually require the mempool to be empty so it not being there should not be a problem. I will open a PR.

fjahr commented:
> > I also removed the mempool and it crashed when reaching validation.cpp:5684. Is it a hard requirement when using assumeutxo that a mempool exists, or is this a bug?
> 
> That appears to be a bug. We actually require the mempool to be empty so it not being there should not be a problem. I will open a PR.

This should work https://github.com/fjahr/bitcoin/commit/21530b489eee6b8db4b9af45430d9d3cb43a459c, but on second thought I am not sure if we need this (yet). Conceptually it makes sense that we change this because, as I said, we don't need the mempool. However, since we don't use the mempool I think this should also not make an impact on the testing speed and even in blocksonly mode there still is a mempool, so this won't be an issue.

This might be getting off-topic so I can open a separate issue to move the discussion.

fjahr commented:
utACK b9ba1a73094f4ad593b527e23d2681f41c225397

sedited reviewed (approved):
ACK b9ba1a73094f4ad593b527e23d2681f41c225397

... but it would be nice to address the nits, will re-ACK quickly :)

mzumsande commented on src/kernel/chainparams.cpp:
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
 
         m_assumeutxo_data = {
-            {
+            {   // For use by unit tests
```
I didn't do that because there are already multiple files that use it (`validation_chainstatemanager_tests`, `validation_chainstate_tests`). Seems less likely that the same would happen for fuzz tests.

mzumsande commented on src/kernel/chainparams.cpp:
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
+                .height = 200,
+                .hash_serialized = AssumeutxoHash{uint256S("0x998b6a337899f04a1cd298ecf0fdc6ab66bfb428f00e5a48998505e91b3b55e0")},
+                .nChainTx = 201,
+                .blockhash = uint256S("0x5e93653318f294fb5aa339d00bbf8cf1c3515488ad99412c37608b139ea63b27,")
```
fixed, the "," went into the hash by mistake (and was ignored). I also fixed the `hash_serialized`, which was wrong somehow.

mzumsande commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -41,8 +41,34 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
+            outfile << Span{metadata};
+        } else {
+            DataStream data_stream{};
+            auto msg_start = Params().MessageStart();
```
fixed, here and at one more spot.

mzumsande commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -73,16 +99,18 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
+            Assert(index);
+            Assert(index->nTx == 0);
+            if (index->nHeight == chainman.GetSnapshotBaseHeight()) {
+                Assert(index->nChainTx != 0);
```
Done!

mzumsande commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -41,8 +41,34 @@ FUZZ_TARGET(utxo_snapshot, .init = initialize_chain)
+            auto msg_start = Params().MessageStart();
+            int base_blockheight{fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 2 * COINBASE_MATURITY)};
+            uint256 base_blockhash{g_chain->at(base_blockheight - 1)->GetHash()};
+            uint64_t m_coins_count{fuzzed_data_provider.ConsumeIntegralInRange<uint64_t>(1, 2 * COINBASE_MATURITY)};
```
Done, use `*3`

mzumsande commented:
Thanks for the reviews! I addressed the feedback by @maflcko and also corrected `hash_serialized` in the chainparams. Not sure why it was wrong before, I'm pretty sure I was able to activate the snapshot before opening the PR on another computer. Probably just an oversight, but maybe someone can confirm that everything is deterministic and the snapshot can be loaded now (e.g. by adding an assert and setting `base_blockheight` and `m_coins_count` to 200 it should crash almost immediately).

> I checked if there is anything we could do in the testing setup to make the test faster. (...)

Sound like good ideas for a follow-up!

sedited reviewed (approved):
ACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

fjahr commented:
utACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

mzumsande commented:
> maybe someone can confirm that everything is deterministic and the snapshot can be loaded now (e.g. by adding an assert and setting `base_blockheight` and `m_coins_count` to 200 it should crash almost immediately).

FYI, confirmed for myself on another computer that this was just an error, and the snapshot can be loaded by the fuzzer.

ryanofsky commented on src/kernel/chainparams.cpp:
```diff
@@ -495,12 +495,19 @@ class CRegTestParams : public CChainParams
             },
+            {
+                // For use by fuzz target src/test/fuzz/utxo_snapshot.cpp
+                .height = 200,
+                .hash_serialized = AssumeutxoHash{uint256S("0x4f34d431c3e482f6b0d67b64609ece3964dc8d7976d02ac68dd7c9c1421738f2")},
+                .nChainTx = 201,
+                .blockhash = uint256S("0x5e93653318f294fb5aa339d00bbf8cf1c3515488ad99412c37608b139ea63b27"),
```
In commit "fuzz: improve utxo_snapshot target" (de71d4dece604907afc4fc26b7788e9c1a4cbecb)

Since fuzz test is not deterministic, and doesn't give us an easy way to know hash values here actually allow snapshot to be loaded, it might be good to write a separate regtest generating a an empty chain with height 200 and verifying these hashes.

ryanofsky reviewed (commented):
Post-merge code review ACK de71d4dece604907afc4fc26b7788e9c1a4cbecb

This seems like a good change in principle, but I'm not actually sure what additional test coverage it provides. The only valid snapshot that can be loaded seems to be one that is hardcoded for a chain with height 200 and no transactions except for coinbases. After the snapshot is generated there is only a single `ConsumeBool` call determining whether or not headers for the blocks are added before loading the snapshot.

So basically it creates a hardcoded snapshot, and then tests loading it with and without headers. I could be missing something, but it seems like that is not really taking advantage of the fuzzer or testing something that would be not more convenient to run as a unit test.

It would probably open up a lot more flexibility for the fuzz test if it were able to modify the `m_assumeutxo_data` vector instead of relying on hardcoded values.

mzumsande commented:
> This seems like a good change in principle, but I'm not actually sure what additional test coverage it provides. The only valid snapshot that can be loaded seems to be one that is hardcoded for a chain with height 200 and no transactions except for coinbases. After the snapshot is generated there is only a single `ConsumeBool` call determining whether or not headers for the blocks are added before loading the snapshot.

Note that loading a valid snapshot is not the only goal of this PR. It also allows for a better coverage of `ActivateSnapshot()` which was previously hard to reach because the fuzzer already had a hard time fuzzing valid metadata (after #29612) and would abort without it. E.g. it has zero coverage in https://maflcko.github.io/b-c-cov/fuzz.coverage/src/validation.cpp.gcov.html though I'm not sure how up-do-date the corpus is, and fuzzing locally I also didn't reach `ActivateSnapshot()` (maybe I wasn't patient enough).

I think for more interesting fuzzing we'd have to think about allowing dynamic registration of snapshots in tests (or enabling an option to just skip the check for the hash in chainparams). 

Some time ago, I thought about this in the context of functional tests (to avoid having to edit chainparams whenever we change something in a functional test), but after realizing that the registered snapshot data from chainparams is not only necessary during registration but also in a restart (`nChainTx`) it seemed too crude to me and I didn't open a PR. Might make more sense for fuzzing though.
